### ISSUE TITLE
Feature Request: Cap color temp

### ISSUE THREAD

**Issue by KlaraJa** [basnijholt/adaptive-lighting#989]
Hi,

I would be a nice feature, if I could cap the color temperature at e.g. 3400K or something else.

Because I have a room where I want to have a nice warm color at night and a bit colder over the day but don't want it to only reach that color at the peak of the day.

A bit like this graph (normal curve of color changes at day) but where the straight line is the cap at 3400K.

<img width="1008" alt="Untitled" src="https://github.com/basnijholt/adaptive-lighting/assets/111448619/110e92fc-520e-479d-9fdd-bdf88ec9e16b">

**Comment by ThinkChaos** [comment]
IIUC that's what `min_color_temp` and `max_color_temp` do?

**Comment by KlaraJa** [comment]
> IIUC that's what `min_color_temp` and `max_color_temp` do?

No, they are just changing the min/max position of the curve and don't make a cap below/above the turning point of the curve

**Comment by lordratner** [comment]
To make the issue simpler, the color temp should have the same curve options as the brightness. I'm actually kind of bummed to learn we can't tune that, because making my lights go fully-warm by sunset is terrible behavior. I want the color temp to transition while the brightness is transitioning, not during the daylight hours.

**Comment by lordratner** [comment]
Ok, looking at the code I think this is already (mostly) possible. Here's how:

You need to first set the min and max color temps to what you want during the day **only**. So whatever color temp you want at sunset, set that as the min. If you want a single color temp from sunrise to sunset, use the same min and max temp. 

Now, set the sleep temp to something really low. You can even go lower than what the light can handle to adjust this. Make sure you have transition_until_sleep enabled. This will create a linear fade of the color temp starting at your min_temp setting at sunset and ending at your sleep_temp setting at solar midnight. It will ramp back up along the same rate until sunrise. 

If you set your sleep_temp to the minimum temp of the light, it's not going to get there until solar midnight. If you want to reach that warmth level earlier, set the sleep_temp **below** what the light can handle, so it gets to the actual minimum earlier in the night.

**Comment by bkeys818** [comment]
I'd like to second this request.

As @lordratner mentioned, it's possible to configure things to get somewhat of the desired function. The issue is that then you lose the ability to reach a lower bulb temperature in the night and hold it there.

In a perfect world, I'd be able to have the color temperature look something like this graph.

<img width="788" alt="Image" src="https://github.com/user-attachments/assets/1b9fc04d-09cc-48cc-b5c2-f8ca8a9cc6ad" />

**Comment by KlaraJa** [comment]
@bkeys818 That is exactly what I meant in the first place.
Your graph explains it a lot better, thanks.

**Comment by bkeys818** [comment]
I ended doing what @lordratner suggested and adding an automation to turn off adaptive lighting at 2500k and then turn it back on at 4am.

**Comment by basnijholt** [comment]
@bkeys818, do you have a closed form equation for that? If so it could be added as a function for setting the shape of the curve.

**Comment by bkeys818** [comment]
@basnijholt, I actually used GraphJS to create it. I was trying to use the graph to convey more of a general idea, nothing exact.

**Comment by bkeys818** [comment]
Come to think of it, what I want is simply for the color temperature to following the brightness graph.

**Comment by lordratner** [comment]
> [@bkeys818](https://github.com/bkeys818), do you have a closed form equation for that? If so it could be implemented as a function for setting the shape of the curve.

So after looking at the code, I think what you're doing now is coming up with a y-axis measurement of the solar position in the sun_position function, then using that position (between 0 and 1 for min/max color temp and between 0 and -1 for min/sleep color temp) in the color_temp_kelvin function.

Under that construct, you could create an offset that would surround the solar noon and midnight positions, then modify this formula:         `return k * (1 - ((target_ts - h) / (h - x)) ** 2)` such that instead of h, it would be h adjusted for the offset. 

I think that's a really complicated way of doing it, personally. I think the best answer is to have color temp set up exactly how brightness is, but independently. 

So, have the section for brightness and all the features as-is. Put that all under the section title of "brightness"

Then have a section called "color temp" that has the same options (with a few differences required to cover rgb vs kelvin). The use gets to decide exactly how they want brightness to scale, and how they want color temp to scale. And as a bonus, all the math is already in there.

Here's what the config would look like in my head:

------------------------------------------------------------------------
**General Settings**

- lights: List of light entity_ids to be controlled (may be empty). 🌟
- interval
- transition
- initial_transition
- sleep_transition
- sunrise_time
- min_sunrise_time
- max_sunrise_time
- sunrise_offset
- sunset_time
- min_sunset_time
- max_sunset_time
- sunset_offset


**Brightness**

- min_brightness slider
- max_brightness slider
- sleep brightness
- brightness mode
- brightness_mode_time_dark
- brightness_mode_time_light

**Color Temp**

- min_color_temp
- max_color_temp
- prefer_rgb_color
- sleep_rgb_or_color_temp
- sleep_color_temp
- sleep_rgb_color
- color_temp_mode  <-- Added (same options as brightness_mode)
- color_temp_mode_time_dark  <-- Added 
- color_temp_mode_time_light  <-- Added 

**Advanced**

- take_over_control
- detect_non_ha_changes
- autoreset_control_seconds
- only_once
- adapt_only_bare_turn_on
- separate_turn_on_commands
- send_split_delay
- adapt_delay
- skip_redundant_commands
- intercept
- multi_light_intercept
- include_config_in_attributes

---------------------------------------------------

This gives the user a lot more control over color temp, without changing the basic theory of the integration. It also makes both brightness and color temp the same in terms of how the user thinks about controlling them. 

What do you think, @bkeys818 @basnijholt ?

**Comment by basnijholt** [comment]
If at all possible, I would rather not have more parameters. Online, I have read feedback where people say not to use adaptive lighting because it is too complicated with its many options.

**Comment by lordratner** [comment]
> If at all possible, I would rather not have more parameters. Online, I have read feedback where people say not to use adaptive lighting because it is too complicated with its many options.

So with what I posted above, the brightness section at "default" would require no inputs, and the transition_until_sleep setting is removed. 

And at least this way, there are a couple more settings for color_temp, but they are the same settings that were used for brightness, making the setup more unified... I think.

**Comment by lordratner** [comment]
I think that breaking the configuration into sections, and putting the necessary settings at the top, and all the fine-tuning at the bottom (with defaults set) would make it very easy for a new user who just wants adaptive lighting to work. Here's another suggestions for how to arrange the configuration to make it even easier for the beginner user. They would only have to change the "General Settings" section to get the integration running.

------------------------------
**General Settings**

- lights: List of light entity_ids to be controlled (may be empty). 🌟
- min_brightness slider
- max_brightness slider
- sleep brightness
- min_color_temp
- max_color_temp
- sleep_color_temp


**Advanced**
- sunrise_time
- min_sunrise_time
- max_sunrise_time
- sunrise_offset
- sunset_time
- min_sunset_time
- max_sunset_time
- sunset_offset

**Brightness Advanced**
- brightness mode
- brightness_mode_time_dark
- brightness_mode_time_light

**Color Temp Advanced**
- color_temp_mode  <-- Added (same options as brightness_mode)
- color_temp_mode_time_dark  <-- Added 
- color_temp_mode_time_light  <-- Added 
- prefer_rgb_color
- sleep_rgb_or_color_temp
- sleep_rgb_color


**Configuration**
- interval
- transition
- initial_transition
- sleep_transition
- take_over_control
- detect_non_ha_changes
- autoreset_control_seconds
- only_once
- adapt_only_bare_turn_on
- separate_turn_on_commands
- send_split_delay
- adapt_delay
- skip_redundant_commands
- intercept
- multi_light_intercept
- include_config_in_attributes

**Comment by bkeys818** [comment]
> If at all possible, I would rather not have more parameters. Online, I have read feedback where people say not to use adaptive lighting because it is too complicated with its many options.

I second that. I'm guessing that making the configuration view dynamic isn't possible. Is it based of a yaml file?

**Comment by RoboMagus** [comment]
> If at all possible, I would rather not have more parameters. Online, I have read feedback where people say not to use adaptive lighting because it is too complicated with its many options.

I get where this is coming from, but not adding functionality because people that might not even use it find it hard to understand should not set the norm. The extent to which this integration can be configured to everyones liking is its major strength in my opinion. Not a downside! 

I fully agree with the suggestion made by @lordratner that properly sectioning the parameters provides the best for every user regardless of experience level. Hiding the more _advanced_ configuration keeps the less experienced users from feeling intimidated while keeping them available for the people that would like to make use of these features. This way everybody wins! :)
